### PR TITLE
Workaround ROCm global pool

### LINKED ISSUE #15482: [ROCm] Global (average) Pooling unusable.
### Describe the issue

1. Crash on some shapes
2. Incorrect result on some shape

## To reproduce a crash 
Run the following single node model
```python
import numpy as np
import onnx
import onnxruntime as ort

batch=1
channel=64
dim1 = 410
dim2 = 400

ort.set_default_logger_severity(0)
ort.set_default_logger_verbosity(1000)

x = onnx.helper.make_tensor_value_info("x", onnx.TensorProto.FLOAT16, [batch, channel, dim1, dim2])
y = onnx.helper.make_tensor_value_info("y", onnx.TensorProto.FLOAT16, [batch, channel, 1, 1])

node = onnx.helper.make_node("GlobalAveragePool", inputs=["x"], outputs=["y"])
graph = onnx.helper.make_graph([node], "GP", [x], [y])
model = onnx.helper.make_model(graph)

sess = ort.InferenceSession(
    model.SerializeToString(), providers=[("ROCMExecutionProvider", {"miopen_conv_use_max_workspace": False})]
)

x = np.random.randn(batch, channel, dim1, dim2).astype(np.float16)
sess.run(input_feed = {"x": x}, output_names = ["y"])

```
will create the following error:
```
MIOpen(HIP): Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_CODEGEN_BC_TO_RELOCATABLE: ERROR (1)
MIOpen(HIP): Error [BuildOcl] comgr status = ERROR (1)
MIOpen(HIP): Warning [BuildOcl] error: stack frame size (328004) exceeds limit (131056) in function 'mloPoolingG'
1 error generated.

MIOpen Error: /long_pathname_so_that_rpms_can_package_the_debug_info/data/driver/MLOpen/src/hipoc/hipoc_program.cpp:304: Code object build failed. Source: MIOpenPooling.cl
2023-04-12 08:27:32.942631957 [E:onnxruntime:Default, rocm_call.cc:119 RocmCall] MIOPEN failure 7: miopenStatusUnknownError ; GPU=0 ; hostname=linmif39a00000F ; file=/home/guangyunhan/onnxruntime/onnxruntime/core/providers/rocm/nn/pool.cc ; line=226 ; expr=PoolingForwardHelper(GetMiopenHandle(context), pooling_desc, &alpha, x_tensor, x_data, &beta, y_tensor, y_data); 
2023-04-12 08:27:32.942678735 [E:onnxruntime:, sequential_executor.cc:516 ExecuteKernel] Non-zero status code returned while running GlobalAveragePool node. Name:'' Status Message: MIOPEN failure 7: miopenStatusUnknownError ; GPU=0 ; hostname=linmif39a00000F ; file=/home/guangyunhan/onnxruntime/onnxruntime/core/providers/rocm/nn/pool.cc ; line=226 ; expr=PoolingForwardHelper(GetMiopenHandle(context), pooling_desc, &alpha, x_tensor, x_data, &beta, y_tensor, y_data);
```

The maxmium shape it can run is
```python 
batch=1
channel=64
dim1 = 255
dim2 = 255
```

This fix the problem by switching the global pool to use reduction instead.

This problem impact the usability of ROCm EP of our internal users.

## To reproduce an incorrect result

```python
import numpy as np
import onnx
import onnxruntime as ort

batch=1
channel=3
dim1 = 255
dim2 = 255

ort.set_default_logger_severity(0)
ort.set_default_logger_verbosity(1000)

x = onnx.helper.make_tensor_value_info("x", onnx.TensorProto.FLOAT16, [batch, channel, dim1, dim2])
y = onnx.helper.make_tensor_value_info("y", onnx.TensorProto.FLOAT16, [batch, channel, 1, 1])

node = onnx.helper.make_node("GlobalAveragePool", inputs=["x"], outputs=["y"])
graph = onnx.helper.make_graph([node], "GP", [x], [y])
model = onnx.helper.make_model(graph)

x = np.random.uniform(low=0.0, high=1.10, size=(batch, channel, dim1, dim2)).astype(np.float16)

sess = ort.InferenceSession(model.SerializeToString(), providers=["CPUExecutionProvider"])
ref = sess.run(input_feed = {"x": x}, output_names = ["y"])[0]

sess = ort.InferenceSession(
    model.SerializeToString(), providers=[("ROCMExecutionProvider", {"miopen_conv_use_max_workspace": False})]
)
y = sess.run(input_feed = {"x": x}, output_names = ["y"])[0]

print(ref.shape)
print(y.shape)
print(ref)
print(y)
print("max relative error:", np.abs((ref-y)/ref).max())
```

### Urgency

Must be fixed

### Platform

Linux

### OS Version

no apply

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

d49a8de

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Other / Unknown

### Execution Provider Library Version

ROCm 5.4.2

### PR DESCRIPTION
**Pull request by cloudhan**
Fix #15482

This also improve the perf for it, comparison is meanless tho, the original impl is incorrect.
| shape | Original (numerically incorrect) | New (numerically correct) |
| ------- | --- | --- |
|1,64,255,255  | 0.16797351837158203 | 0.0012824535369873047
|32,64,255,255 | 1.3541345596313477  | 0.008842945098876953
|32,64,14,14  | 0.00106048583984375  | 0.0008864402770996094

### CODE REVIEW

File: onnxruntime/core/providers/rocm/nn/pool.cc [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+  } else {
+    ORT_NOT_IMPLEMENTED();
+  }
+
+  miopenDataType_t miopen_type_X = MiopenTensor::GetDataType<HipT>();
+
+  MiopenReduceDescriptor reduce_desc;
+  if constexpr (std::is_same<T, MLFloat16>::value) {
+    ORT_RETURN_IF_ERROR(reduce_desc.Set(
+        reduce_op, MiopenTensor::GetDataType<float>(), MIOPEN_REDUCE_TENSOR_FLATTENED_INDICES));
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/nn/pool.cc:311]: Why use `MiopenTensor::GetDataType<float>()` here?

File: onnxruntime/core/providers/rocm/nn/pool.cc [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  } else {
+    ORT_RETURN_IF_ERROR(reduce_desc.Set(reduce_op, miopen_type_X, MIOPEN_REDUCE_TENSOR_FLATTENED_INDICES));
+  }
+
+  ORT_ENFORCE(x_shape.NumDimensions() >= 3);
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/nn/pool.cc:316]: Similar to Line291? Maybe only need to keep the judgement once.

File: onnxruntime/core/providers/rocm/nn/pool.cc [hunk 3]
```diff
@@ -0,0 +1,353 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared_library/provider_api.h"
+#include "core/providers/rocm/miopen_common.h"
+#include "core/providers/rocm/nn/pool.h"
+#include "core/providers/rocm/nn/max_pool_with_index.h"
+#include "core/providers/rocm/math/unary_elementwise_ops_impl.h"
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/nn/pool.cc:8]: Could you please sort the header file?

File: onnxruntime/core/providers/rocm/nn/pool.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/providers/rocm/rocm_kernel.h"
+#include "core/providers/rocm/miopen_common.h"
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/nn/pool.h:7]: Same here. Could you please sort the header file?

File: onnxruntime/core/providers/rocm/nn/pool.cc (hunk 1 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/nn/pool.cc:311]: Copied from Reduce* Op impl. It should just use acc type as fp32 for numerical stability.

File: onnxruntime/core/providers/rocm/nn/pool.cc (hunk 2 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/nn/pool.cc:316]: Great catch

File: onnxruntime/core/providers/rocm/nn/pool.cc [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  auto kernel_shape = pool_attrs_.kernel_shape;
+  auto pads = pool_attrs_.pads;
+  auto strides = pool_attrs_.strides;
+
+  if (pool_attrs_.global_pooling) {
+    ORT_ENFORCE("unreachable");
```
Comment (kailums) [inline onnxruntime/core/providers/rocm/nn/pool.cc:172]: ORT_ENFORCE need a condition. may be ORT_ENFORCE(!pool_attrs_.global_pooling, '....')?

File: onnxruntime/core/providers/rocm/nn/pool.cc [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  auto kernel_shape = this->pool_attrs_.kernel_shape;
+  auto pads = this->pool_attrs_.pads;
+  auto strides = this->pool_attrs_.strides;
+
+  if (this->pool_attrs_.global_pooling) {
+    ORT_ENFORCE("unreachable");
```
Comment (kailums) [inline onnxruntime/core/providers/rocm/nn/pool.cc:251]: same as line 172

File: onnxruntime/core/providers/rocm/nn/pool.cc (hunk 3 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/nn/pool.cc:7]: No, move the header will cause additional compilation errors to pop up. It seem some of the included headers are not self contained due to historical issue. I don't want to touch them headers to fix it.

**Review by PeixuanZuo — DISMISSED** [review]
LGTM, Thanks.

**Comment by kailums** [comment]
approve

**Review by kailums — APPROVED** [review]
approve

**Review by kailums — APPROVED** [review]

**Pull request merged**